controllers/serviceaccount_controller.go

```diff
@@ -249,74 +255,81 @@ func (r ServiceAccountReconciler) ReconcileNormal(guestClusterCtx *vmwarecontext
 		}
 	}()
 
-	pSvcAccounts, err := getProviderServiceAccounts(guestClusterCtx.ClusterContext)
+	pSvcAccounts, err := r.getProviderServiceAccounts(ctx, guestClusterCtx.ClusterContext)
 	if err != nil {
-		guestClusterCtx.Logger.Error(err, "Error fetching provider serviceaccounts")
+		log.Error(err, "Error fetching ProviderServiceAccounts")
 		return reconcile.Result{}, err
 	}
-	err = r.ensureProviderServiceAccounts(guestClusterCtx, pSvcAccounts)
+	err = r.ensureProviderServiceAccounts(ctx, guestClusterCtx, pSvcAccounts)
 	if err != nil {
-		guestClusterCtx.Logger.Error(err, "Error ensuring provider serviceaccounts")
+		log.Error(err, "Error ensuring ProviderServiceAccounts")
 		return reconcile.Result{}, err
 	}
 
 	return reconcile.Result{}, nil
 }
 
 // Ensure service accounts from provider spec is created.
-func (r ServiceAccountReconciler) ensureProviderServiceAccounts(guestClusterCtx *vmwarecontext.GuestClusterContext, pSvcAccounts []vmwarev1.ProviderServiceAccount) error {
+func (r *ServiceAccountReconciler) ensureProviderServiceAccounts(ctx context.Context, guestClusterCtx *vmwarecontext.GuestClusterContext, pSvcAccounts []vmwarev1.ProviderServiceAccount) error {
+	log := ctrl.LoggerFrom(ctx)
+
 	for i, pSvcAccount := range pSvcAccounts {
+		// Note: We have to use := here to not overwrite log & ctx outside the for loop.
+		log := log.WithValues("ProviderServiceAccount", klog.KRef(pSvcAccount.Namespace, pSvcAccount.Name))
```

Comment:
Nit: Might be less confusing if we use new variable names here instead of shadowing.

```suggestion
		log := log.WithValues("ProviderServiceAccount", klog.KRef(pSvcAccount.Namespace, pSvcAccount.Name))
```

Reply:
I prefer shadowing. If we don't shadow it's very easy to accidentally use the wrong log or context.

We intentionally shadow log & ctx here so that within the scope of the for loop the shadowed log/ctx has to be used

Reply:
I think the main difficulty here is deciding when we have to use `:=` over `=`, but we would have the same problem if in the cases where we currently use `:=` we would use a different variable name. Once we made the right call we're good.